climada/engine/impact.py

```diff
@@ -48,6 +47,410 @@
+
+    def impact(self, save_mat=True):
+        """Compute the impact of a hazard on exposures.
+
+        Parameters
+        ----------
+        save_mat : bool
+            if true, save the total impact matrix (events x exposures)
+
+        Examples
+        --------
+            >>> haz = Hazard.from_mat(HAZ_DEMO_MAT)  # Set hazard
+            >>> impfset = ImpactFuncSet.from_excel(ENT_TEMPLATE_XLS) # Set impact functions
+            >>> exp = Exposures(pd.read_excel(ENT_TEMPLATE_XLS)) # Set exposures
+            >>> impcalc = ImpactCal(exp, impfset, haz)
+            >>> imp = impcalc.insured_impact()
+            >>> imp.aai_agg
+
+        Note
+        ----
+        Deductible and/or cover values in the exposures are ignored.
+        """
+        impf_col = self.exposures.get_impf_column(self.hazard.haz_type)
+        exp_gdf = self.minimal_exp_gdf(impf_col)
+        LOGGER.info('Calculating impact for %s assets (>0) and %s events.',
+                    self.n_events, self.n_events)
+        imp_mat_gen = self.imp_mat_gen(exp_gdf, impf_col)
+        if save_mat:
+            self.imp_mat = self.stitch_impact_matrix(imp_mat_gen)
+            at_event, eai_exp, aai_agg = self.risk_metrics()
+        else:
+            at_event, eai_exp, aai_agg = self.stitch_risk_metrics(imp_mat_gen)
+        return Impact.from_eih(
+            self.exposures, self.impfset, self.hazard,
+            at_event, eai_exp, aai_agg, self.imp_mat
+            )
+
+    def insured_impact(self, save_mat=False):
+        """Compute the impact of a hazard on exposures with a deductible and/or
+        cover.
+
+        For each exposure point, the impact per event is obtained by
+        substracting the deductible (and is maximally equal to the cover).
+
+        Parameters
+        ----------
+        save_mat : bool
+            if true, save the total impact matrix (events x exposures)
+
+        Examples
+        --------
+            >>> haz = Hazard.from_mat(HAZ_DEMO_MAT)  # Set hazard
+            >>> impfset = ImpactFuncSet.from_excel(ENT_TEMPLATE_XLS) # Set impact functions
+            >>> exp = Exposures(pd.read_excel(ENT_TEMPLATE_XLS)) # Set exposures
+            >>> impcalc = ImpactCal(exp, impfset, haz)
+            >>> imp = impcalc.insured_impact()
+            >>> imp.aai_agg
+
+        See also
+        --------
+        apply_deductible_to_mat:
+            apply deductible to impact matrix
+        apply_cover_to_mat:
+            apply cover to impact matrix
+        """
+        impf_col = self.exposures.get_impf_column(self.hazard.haz_type)
+        exp_gdf = self.minimal_exp_gdf(impf_col)
+        LOGGER.info('Calculating impact for %s assets (>0) and %s events.',
+                    exp_gdf.size, self.hazard.size)
+
+        if self.imp_mat.size == 0:
+            imp_mat_gen = self.imp_mat_gen(exp_gdf, impf_col)
+        else:
+            imp_mat_gen = ((self.imp_mat, self.exposures.gdf.index.values) for n in range(1))
```

Comment:
I'm not sure if this is on purpose but the definition of the impact matrix (generator) here is different from the one in `impact()`. I don't unterstand the matrix creation well enough to see if this is an issue. Consider specifically the use case where somebody calls the two methods after each other, with different settings for `save_mat`. Why would the calculation change if `self.imp_mat` already exists?

Reply:
I guess I know understand the generator-thing better, but I still have two more questions 😅 

> Thus, if the impact matrix is already computed, we would like to avoid computing it again. 

But shouldn't the same logic apply for the `impact()` method as well? It currently always re-computes the impact matrix, unlike `insured_impact()`.

> The save_mat flag is then simple. Either the ouput Impact object with the insured impact has the imp_mat defined or not. This is independent on whether the imp_mat is defined for the ImpactCalc object.

Note that`save_mat` also determines if the matrix will be stored in the `ImactCalc` object. Is that an over-complication of the interface? If the matrix is computed anyway (and occupies memory during the computation), why not always store (i.e. keep) it?